react/src/App.tsx

```diff
@@ -72,24 +73,46 @@ function BaseApp(props: { darkMode: boolean; toggleDarkMode: () => void }) {
             setAuthenticated(false);
         }
     }
+
     // Check if already signed in
     // Since both apiInfo and a loggedin status are required to render main app, we'll query both in sequence to prevent unnecessary rerenders/reroutes
     useEffect(() => {
         (async () => {
-            const loginResult = await fetch("/api/login", { method: "POST" });
+            const loginResult = await customFetch(`/api/login`, { method: "POST" });
             if (loginResult.ok) {
                 const loginInfo = await loginResult.json();
                 setCurrentUser(loginInfo);
             }
             setAuthenticated(loginResult.ok);
 
-            const apiInfoResult = await fetch("/api");
+            let endpoints: LabSelection[] = [];
+            const availibleEndpoints = await fetch("/labs.json");
+            if (availibleEndpoints.ok) {
+                endpoints = await availibleEndpoints.json();
+            }
+
+            if (endpoints.length > 0) {
+                setAvailableEndpoints(endpoints);
+                return;
+            }
+            localStorage.removeItem("endpoint");
+
+            const apiInfoResult = await fetch(`/api`);
             if (apiInfoResult.ok) {
-                const apiInfo = await apiInfoResult.json();
+                let apiInfo = await apiInfoResult.json();
                 setApiInfo(apiInfo as APIInfo);
             }
         })();
     }, []);
+
+    const setEndpoint = async (newEndpoint: string) => {
```

Comment:
I think this might need to be renamed to something like `setApiInfo`, since the endpoint is actually being set in LabDropdownSelect.tsx. It also might be worth using this function in place of the code lines 100-104 if `customFetch` can be called at this point (since the endpoint has already been set in `LabDropdownSelect`).

Reply:
Fixed it